File: src/utils/recipeTags.js
import { canonicalIngredient } from '../data/ingredientKnowledge'
import { RECIPE_TAGS } from '../data/recipeStyles'

const MEAT_FISH_CANON = new Set([
  'chicken', 'beef', 'meat', 'fish', 'salmon', 'tuna', 'turkey', 'lamb', 'pork', 'steak', 'ground beef',
])

const DAIRY_EGG_CANON = new Set([
  'milk', 'egg', 'eggs', 'cheese', 'butter', 'cream', 'yogurt', 'ricotta', 'parmesan', 'feta',
])

const HONEY_CANON = new Set(['honey'])

const GLUTEN_CANON = new Set(['flour', 'pasta', 'bread', 'wheat', 'noodles', 'tortilla'])

const SPICY_CANON = new Set(['chili', 'pepper flakes', 'hot sauce', 'curry powder'])

function ingredientCanons(recipe) {
  return (recipe.ingredients ?? [])
    .map((item) => canonicalIngredient(String(item)))
    .filter(Boolean)
}

/**
 * Derive dietary/classification tags from nutrition, ingredients, and context.
 */
export function deriveRecipeTags(
  recipe,
  {
    category = 'dairy',
    isGlutenFree = false,
    recipeType = 'meal',
    spiceLevel = 0,
    cookTime = 30,
  } = {},
) {
  const tags = new Set()
  const nutrition = recipe.nutrition ?? {}
  const calories = nutrition.calories ?? 0
  const protein = nutrition.protein ?? 0
  const fat = nutrition.fat ?? 0
  const sugar = nutrition.sugar ?? nutrition.sugars ?? 0
  const healthScore = recipe.healthScore ?? nutrition.healthScore ?? 50
  const canons = ingredientCanons(recipe)
  const hasMeatFish = canons.some((item) => MEAT_FISH_CANON.has(item))
  const hasDairyEgg = canons.some((item) => DAIRY_EGG_CANON.has(item))
  const hasHoney = canons.some((item) => HONEY_CANON.has(item))
  const hasGlutenIng = canons.some((item) => GLUTEN_CANON.has(item))
  const hasSpicyIng = canons.some((item) => SPICY_CANON.has(item))

  if (RECIPE_TAGS.vegetarian.categories.includes(category) && !hasMeatFish) {
    tags.add('vegetarian')
  }
  if (!hasMeatFish && !hasDairyEgg && !hasHoney) {
    tags.add('vegan')
  }
  if (isGlutenFree && !hasGlutenIng) {
    tags.add('glutenFree')
  }
  if (protein >= RECIPE_TAGS.highProtein.minProtein) {
    tags.add('highProtein')
  }
  if (cookTime <= RECIPE_TAGS.quick.maxTime) {
    tags.add('quick')
  }

  const isIndulgentDessert = recipeType === 'dessert' && calories > 420
  const isHighCalorie = calories > 520 || fat > 28

  if (
    !isIndulgentDessert &&
    healthScore >= RECIPE_TAGS.healthy.minHealthScore &&
    calories <= 550 &&
    fat <= 24 &&
    sugar <= 22
  ) {
    tags.add('healthy')
  }

  if (
    !isIndulgentDessert &&
    calories <= 420 &&
    fat <= 18 &&
    healthScore >= 70 &&
    sugar <= 18
  ) {
    tags.add('dietFriendly')
  }

  if (
    recipeType === 'meal' &&
    protein >= 28 &&
    calories <= 580 &&
    fat <= 22 &&
    sugar <= 25
  ) {
    tags.add('postWorkout')
  }

  if (
    spiceLevel === 0 &&
    !hasSpicyIng &&
    calories <= 650 &&
    recipeType !== 'dessert' &&
    sugar <= 20
  ) {
    tags.add('childFriendly')
  }

  if (isHighCalorie || healthScore < 62) {
    tags.add('comfortFood')
  }

  if (isIndulgentDessert) {
    tags.delete('healthy')
    tags.delete('dietFriendly')
    tags.delete('childFriendly')
    tags.delete('postWorkout')
  }

  if (spiceLevel >= 2 || hasSpicyIng) {
    tags.delete('childFriendly')
  }

  if (hasMeatFish) {
    tags.delete('vegetarian')
    tags.delete('vegan')
  } else if (hasDairyEgg || hasHoney) {
    tags.delete('vegan')
  }

  if (!isGlutenFree || hasGlutenIng) {
    tags.delete('glutenFree')
  }

  return [...tags]
}

/**
 * Ensure declared tags do not contradict derived facts.
 */
export function validateRecipeTags(recipe, derivedTags) {
  const derived = new Set(derivedTags)
  const guarded = new Set([
    'healthy',
    'dietFriendly',
    'childFriendly',
    'postWorkout',
    'vegetarian',
    'vegan',
    'glutenFree',
    'comfortFood',
    'highProtein',
  ])
  const invalid = (recipe.tags ?? []).filter((tag) => guarded.has(tag) && !derived.has(tag))
  return {
    ok: invalid.length === 0,
    invalid,
    derivedTags,
  }
}

export function applyDerivedRecipeTags(recipe, context = {}) {
  const derivedTags = deriveRecipeTags(recipe, context)
  return {
    ...recipe,
    tags: derivedTags,
  }
}

File: backend/recipe_tags.py
"""Derive recipe tags from nutrition and ingredients — never trust model tags blindly."""

from __future__ import annotations

from ingredient_relevance import canonical_ingredient
from measurement_units import strip_quantity_prefix

MEAT_FISH_CANON = frozenset(
    {"chicken", "beef", "fish", "salmon", "tuna", "turkey", "lamb", "pork", "meat", "steak"}
)
DAIRY_EGG_CANON = frozenset(
    {"milk", "egg", "eggs", "cheese", "butter", "cream", "yogurt", "ricotta", "parmesan", "feta"}
)
GLUTEN_CANON = frozenset({"flour", "pasta", "bread", "wheat", "noodles", "tortilla"})
SPICY_CANON = frozenset({"chili", "pepper flakes", "hot sauce", "curry powder"})

HIGH_PROTEIN_MIN = 25
HEALTHY_MIN_SCORE = 80
HEALTHY_MAX_CAL = 550
HEALTHY_MAX_FAT = 24
DIET_MAX_CAL = 420
DIET_MAX_FAT = 18
DIET_MIN_SCORE = 70
DIET_MAX_SUGAR = 18
CHILD_MAX_CAL = 650
POST_WORKOUT_MIN_PROTEIN = 28
POST_WORKOUT_MAX_CAL = 580
POST_WORKOUT_MAX_FAT = 22
QUICK_MAX_TIME = 25


def _ingredient_canons(recipe: dict) -> list[str]:
    result = []
    for item in recipe.get("ingredients") or []:
        canon = canonical_ingredient(strip_quantity_prefix(str(item)))
        if canon:
            result.append(canon)
    return result


def derive_recipe_tags(
    recipe: dict,
    *,
    category: str = "dairy",
    is_gluten_free: bool = False,
    recipe_type: str = "meal",
    spice_level: int = 0,
    cook_time: int = 30,
) -> list[str]:
    tags: set[str] = set()
    nutrition = recipe.get("nutrition") or {}
    calories = float(nutrition.get("calories") or 0)
    protein = float(nutrition.get("protein") or 0)
    fat = float(nutrition.get("fat") or 0)
    sugar = float(nutrition.get("sugar") or nutrition.get("sugars") or 0)
    health_score = float(recipe.get("healthScore") or nutrition.get("healthScore") or 50)

    canons = _ingredient_canons(recipe)
    canon_set = set(canons)
    has_meat_fish = bool(canon_set & MEAT_FISH_CANON)
    has_dairy_egg = bool(canon_set & DAIRY_EGG_CANON)
    has_gluten = bool(canon_set & GLUTEN_CANON)
    has_spicy = bool(canon_set & SPICY_CANON)

    if category in ("dairy", "parve", "vegan") and not has_meat_fish:
        tags.add("vegetarian")
    if not has_meat_fish and not has_dairy_egg:
        tags.add("vegan")

    if is_gluten_free and not has_gluten:
        tags.add("glutenFree")

    if protein >= HIGH_PROTEIN_MIN:
        tags.add("highProtein")

    if cook_time <= QUICK_MAX_TIME:
        tags.add("quick")

    is_indulgent_dessert = recipe_type == "dessert" and calories > 420
    is_high_calorie = calories > 520 or fat > 28

    if (
        not is_indulgent_dessert
        and health_score >= HEALTHY_MIN_SCORE
        and calories <= HEALTHY_MAX_CAL
        and fat <= HEALTHY_MAX_FAT
        and sugar <= 22
    ):
        tags.add("healthy")

    if (
        not is_indulgent_dessert
        and calories <= DIET_MAX_CAL
        and fat <= DIET_MAX_FAT
        and health_score >= DIET_MIN_SCORE
        and sugar <= DIET_MAX_SUGAR
    ):
        tags.add("dietFriendly")

    if (
        recipe_type == "meal"
        and protein >= POST_WORKOUT_MIN_PROTEIN
        and calories <= POST_WORKOUT_MAX_CAL
        and fat <= POST_WORKOUT_MAX_FAT
        and sugar <= 25
    ):
        tags.add("postWorkout")

    if (
        spice_level == 0
        and not has_spicy
        and calories <= CHILD_MAX_CAL
        and recipe_type != "dessert"
        and sugar <= 20
    ):
        tags.add("childFriendly")

    if is_high_calorie or health_score < 62:
        tags.add("comfortFood")

    if is_indulgent_dessert:
        tags.discard("healthy")
        tags.discard("dietFriendly")
        tags.discard("childFriendly")
        tags.discard("postWorkout")

    if spice_level >= 2 or has_spicy:
        tags.discard("childFriendly")

    if has_meat_fish:
        tags.discard("vegetarian")
        tags.discard("vegan")
    elif has_dairy_egg:
        tags.discard("vegan")

    if not is_gluten_free or has_gluten:
        tags.discard("glutenFree")

    return sorted(tags)


def apply_derived_recipe_tags(recipe: dict, **context) -> dict:
    return {**recipe, "tags": derive_recipe_tags(recipe, **context)}
